internal/pkg/api/metrics.go

```diff
@@ -37,83 +43,149 @@ var (
-	rt.bodyOut = monitoring.NewUint(registry, "body_out")
+func newGauge(registry *metricsRegistry, name string) *statsGauge {
+	g := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: registry.fullName,
+		Name:      name,
+	})
+	registry.promReg.MustRegister(g)
+	return &statsGauge{
+		metric: monitoring.NewUint(registry.registry, name),
+		gauge:  g,
+	}
 }
 
-func init() {
-	err := report.SetupMetrics(logger.NewZapStub("instance-metrics"), build.ServiceName, version.DefaultVersion)
-	if err != nil {
-		log.Error().Err(err).Msg("unable to initialize metrics")
+func (g *statsGauge) Add(delta uint64) {
+	g.metric.Add(delta)
+	g.gauge.Add(float64(delta))
+}
+
+func (g *statsGauge) Inc() {
+	g.metric.Inc()
+	g.gauge.Inc()
+}
+
+func (g *statsGauge) Dec() {
+	g.metric.Dec()
+	g.gauge.Dec()
+}
+
+// statsCounter wraps counters for internal libbeat and prometheus
+type statsCounter struct {
+	metric  *monitoring.Uint
+	counter prometheus.Counter
+}
+
+func newCounter(registry *metricsRegistry, name string) *statsCounter {
+	c := prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: registry.fullName,
+		Name:      name,
+	})
+	registry.promReg.MustRegister(c)
+	return &statsCounter{
+		metric:  monitoring.NewUint(registry.registry, name),
+		counter: c,
 	}
+}
 
-	registry = monitoring.Default.NewRegistry("http_server")
-	cntHTTPNew = monitoring.NewUint(registry, "tcp_open")
-	cntHTTPClose = monitoring.NewUint(registry, "tcp_close")
+func (g *statsCounter) Add(delta uint64) {
+	g.metric.Add(delta)
+	g.counter.Add(float64(delta))
+}
 
-	routesRegistry := registry.NewRegistry("routes")
+func (g *statsCounter) Inc() {
+	g.metric.Inc()
+	g.counter.Inc()
+}
 
-	cntCheckin.Register(routesRegistry.NewRegistry("checkin"))
-	cntEnroll.Register(routesRegistry.NewRegistry("enroll"))
-	cntArtifacts.Register(routesRegistry.NewRegistry("artifacts"))
-	cntAcks.Register(routesRegistry.NewRegistry("acks"))
-	cntStatus.Register(routesRegistry.NewRegistry("status"))
-	cntUploadStart.Register(routesRegistry.NewRegistry("uploadStart"))
-	cntUploadChunk.Register(routesRegistry.NewRegistry("uploadChunk"))
-	cntUploadEnd.Register(routesRegistry.NewRegistry("uploadEnd"))
+// routeStats is the generic collection metrics that we collect per API route.
+type routeStats struct {
+	active    *statsGauge
+	total     *statsCounter
+	rateLimit *statsCounter
+	maxLimit  *statsCounter
+	failure   *statsCounter
+	drop      *statsCounter
+	bodyIn    *statsCounter
+	bodyOut   *statsCounter
 }
 
-func (rt *routeStats) IncError(err error) {
+func (rt *routeStats) Register(registry *metricsRegistry) {
+	rt.active = newGauge(registry, "active")
+	rt.total = newCounter(registry, "total_count")
+	rt.rateLimit = newCounter(registry, "error_limit_rate_count")
+	rt.maxLimit = newCounter(registry, "error_limit_max_count")
+	rt.failure = newCounter(registry, "error_fail_count")
+	rt.drop = newCounter(registry, "error_drop_count")
+	rt.bodyIn = newCounter(registry, "body_in_bytes")
```

Comment:
I would try to avoid renaming metrics if possible, as mentioned by @juliaElastic this would be a breaking change. Is there any reason for the change in metric names?
```suggestion
	rt.bodyIn = newCounter(registry, "body_in")
```